test/browser/specs/headless-config-inheritance.test.ts

```diff
@@ -0,0 +1,128 @@
+import { expect, test } from 'vitest'
+import { runInlineTests, ts } from '../../test-utils'
+import { instances, provider } from './utils'
+
+interface HeadlessConfigInheritanceTestCase {
+  name: string
+  rootHeadless?: boolean
+  instanceHeadless?: boolean
+  expectedResolvedInstanceHeadless: boolean
+}
+
+test.each<HeadlessConfigInheritanceTestCase>([
+  {
+    name: 'instance headless is used if root headless is not set',
+    rootHeadless: undefined,
+    instanceHeadless: true,
+    expectedResolvedInstanceHeadless: true,
+  },
+  {
+    name: 'instance headless is used if root headless is not set',
+    rootHeadless: undefined,
+    instanceHeadless: false,
+    expectedResolvedInstanceHeadless: false,
+  },
+  {
+    name: 'root headless is used if instance headless is not set',
+    rootHeadless: true,
+    instanceHeadless: undefined,
+    expectedResolvedInstanceHeadless: true,
+  },
+  {
+    name: 'root headless is used if instance headless is not set',
+    rootHeadless: false,
+    instanceHeadless: undefined,
+    expectedResolvedInstanceHeadless: false,
+  },
+  {
+    name: 'both set to the same value',
+    rootHeadless: true,
+    instanceHeadless: true,
+    expectedResolvedInstanceHeadless: true,
+  },
+  {
+    name: 'both set to the same value',
+    rootHeadless: false,
+    instanceHeadless: false,
+    expectedResolvedInstanceHeadless: false,
+  },
+  {
+    name: 'instance headless overrides root headless',
+    rootHeadless: true,
+    instanceHeadless: false,
+    expectedResolvedInstanceHeadless: false,
+  },
+  {
+    name: 'instance headless overrides root headless',
+    rootHeadless: false,
+    instanceHeadless: true,
+    expectedResolvedInstanceHeadless: true,
+  },
+])(
+  '$name (root=$rootHeadless, instance=$instanceHeadless, expected=$expectedResolvedInstanceHeadless)',
+  async ({
+    rootHeadless,
+    instanceHeadless,
+    expectedResolvedInstanceHeadless,
+  }) => {
+    expect(instances.length).toBeGreaterThan(0)
+
+    const { ctx } = await runInlineTests({
+      'vitest.config.ts': ts`
+        import { fileURLToPath } from 'node:url'
+        import { defineConfig } from 'vitest/config'
+        import instance from './instance.json'
+
+        export default defineConfig({
+          test: {
+            browser: {
+              enabled: true,
+              provider: '${provider}',
+              headless: ${rootHeadless},
+              instances: [
+                instance,
+              ],
+            },
+          },
+          cacheDir: fileURLToPath(new URL('./node_modules/.vite', import.meta.url)),
```

Comment:
`cacheDir` probably not needed because `runInlineTests` will place each attempt inside its own folder